feat: UI context for Max

kappa90 opened the pull request:
> [!IMPORTANT]
> 👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Problem
Max currently has no information about what the user is currently doing in the app.
This PR adds a new UI context logic, that allows us to send additional information to Max.
To start, we support the three insight types already recognized by Max: trends, funnels, retention, as well as dashboards.

## Changes
Frontend:
- New `maxContextLogic` to compile a UI-based context that can be injected in the `api/conversations` endpoint call in `maxThreadLogic`; currently supports insights, dashboards and navigation context (page title, URL path)
- `maxContextLogic` exposes options that can be used by the `taxonomicPopover` to show Max AI context; for now, it only supports one option type, which is "Current page", which allows the user to set whatever is in view into the context (currently only insights and dashboards)
- Update `taxonomyFilterLogic` and `InfiniteList` to support a new `TaxonomicFilterGroupType.MaxAIContext`
- Update `DefinitionPopoverContents` to support `TaxonomicFilterGroupType.MaxAIContext` options and display information about insights/dashboards in view
- Connect `maxContextLogic` to `insightLogic`, `insightDataLogic` and `dashboardLogic` to retrieve data about insights in view and current dashboard
- New `ContextDisplay` and `ContextTags` component to display context pills above Max's text input 
- Use `maxContextLogic` to handle clicks on `TaxonomyPopover` items, correctly mounting and unmounting insights and dashboards logics to load the full objects (the taxonomy popover only sends a barebone object)
- Update `LemonTag` to have a new `xsmall` size, to match the `LemonButton` `xsmall` size
- Update `Popover` to correctly place itself above the button, when the button is at the bottom of the screen and there is no vertical space to display the whole popover content
- Display UI context on `Thread` messages to show which context was injected on each message

Backend:
- `Assistant` class now accepts a `ui_context` object, which is saved on the latest `HumanMessage`, to be retrieved in the conversation history 
- Extract `QueryExecutor`'s query running and formatting logic into separate utils
- New `RootNodeUIContextMixin ` mixin for the `Root` node, which formats `ui_context` into the node's prompt
- Small icon added to the Max's intro description, which shows the current UI context, on hover

New tests:
- New tests for the query formatting and specifically for SQL formatting, as well as for the query runner, which were missing entirely
- New tests for ui_context backend logic
- Tests for all the new frontend logics
- New storybook scene

## Did you write or update any docs for this change?

<!-- Engineers are responsible for doing the first pass at documenting their features and/or code.  -->

- [ ] I've added or updated the docs
- [ ] I've reached out for help from the docs team
- [ ] No docs needed for this change

## How did you test this code?
Added a whole bunch of tests

Commit:
fix: UI context for Max

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Commit:
fix: add storybook scene

Commit:
Merge remote-tracking branch 'origin/master' into max/frontend-context

Commit:
fix: update schema

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Commit:
fix: greptile review

Commit:
Merge remote-tracking branch 'origin/max/frontend-context' into max/frontend-context

kappa90 commented:
I rewrote the entire PR, and edited the original message to show what's been released here.

Commit:
fix: typescript

Commit:
fix: small logic fix

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

Commit:
fix: logic selector

Commit:
fix: tests

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

Commit:
fix: upgrade to modern kea

Commit:
fix: frontend tests

Commit:
Merge remote-tracking branch 'origin/master' into max/frontend-context

Commit:
fix: schema

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 commented on ee/hogai/graph/base.py:
```diff
@@ -101,29 +102,28 @@ def _get_contextual_tools(self, config: RunnableConfig) -> dict[str, Any]:
-            contextual_tools = config["configurable"]["contextual_tools"]
-        except KeyError:
-            return {}
+        contextual_tools = config.get("configurable", {}).get("contextual_tools") or {}
```
`configurable` can be set to `None`, so the line will throw as you can't do `.get` on None.

skoob13 commented on ee/hogai/assistant.py:
```diff
@@ -239,9 +242,25 @@ def _stream(self) -> Generator[str, None, None]:
     def _initial_state(self) -> AssistantState:
         if self._latest_message and self._mode == AssistantMode.ASSISTANT:
-            return AssistantState(messages=[self._latest_message], start_id=self._latest_message.id)
+            # Add ui_context to the message if available
+            if self._ui_context:
+                # remove global_info as it is not added by the user
+                # and we don't want to show it in the client
+                filtered_ui_context = {k: v for k, v in self._ui_context.items() if k not in ["global_info"]}
+                message_with_ui_context = self._latest_message.model_copy(update={"ui_context": filtered_ui_context})
+                return AssistantState(
+                    messages=[message_with_ui_context],
+                    start_id=message_with_ui_context.id,
+                )
```
I don't understand this piece. Why do we pass the UI context twice: in the message and the graph config?

kappa90 replied:
Right, I could find the context from the last message. I'll do that, hadn't thought about it, let me do that.

skoob13 commented on ee/hogai/graph/base.py:
```diff
@@ -25,8 +26,8 @@ def __call__(self, state: AssistantState, config: RunnableConfig) -> PartialAssi
         """
-        thread_id = config["configurable"]["thread_id"]
-        if self._is_conversation_cancelled(thread_id):
+        thread_id = config.get("configurable", {}).get("thread_id")
```
The same as below

skoob13 commented on ee/hogai/graph/base.py:
```diff
@@ -101,29 +102,28 @@ def _get_contextual_tools(self, config: RunnableConfig) -> dict[str, Any]:
+        """
+        Extracts the UI context from the runnable config.
+        """
+        ui_context_data = config.get("configurable", {}).get("ui_context")
```
^

skoob13 commented on ee/hogai/graph/base.py:
```diff
@@ -101,29 +102,28 @@ def _get_contextual_tools(self, config: RunnableConfig) -> dict[str, Any]:
-        except KeyError:
-            return None
-        return distinct_id
+        return config.get("configurable", {}).get("distinct_id") or None
```
^

skoob13 commented on ee/hogai/graph/base.py:
```diff
@@ -101,29 +102,28 @@ def _get_contextual_tools(self, config: RunnableConfig) -> dict[str, Any]:
-            return config["configurable"]["trace_id"]
-        except KeyError:
-            return None
+        return config.get("configurable", {}).get("trace_id") or None
```
^ and here

skoob13 commented on ee/hogai/graph/query_executor/query_runner.py:
```diff
@@ -0,0 +1,197 @@
+from django.core.serializers.json import DjangoJSONEncoder
+
+
+class QueryRunner:
```
Let's rename it to something more meaningful

kappa90 replied:
Renamed to `AssistantQueryExecutor`

skoob13 commented on ee/hogai/graph/root/nodes.py:
```diff
@@ -52,7 +58,163 @@
+
+        # Format standalone insights context
+        insights_context = ""
+        if ui_context.insights:
```
When I attach the current page with a dashboard, it includes both `dashboards.key.insights` and `insights`. I think we have to use their current frontend versions for the dashboards because users might apply ephemeral filters, so I would completely remove the `dashboards.insights` field.

<img width="380" alt="Screenshot 2025-06-05 at 12 09 49" src="https://github.com/user-attachments/assets/21b670c8-5606-46b5-8704-ff3038f4306c" />

kappa90 replied:
This was a bug, insights should have been deduplicated, instead they were sent twice (in a dashboard, and outside a dashboard). Fixed!

skoob13 commented on ee/hogai/graph/root/nodes.py:
```diff
@@ -52,7 +58,163 @@
+                dashboard_contexts.append(dashboard_text)
+
+            if dashboard_contexts:
+                dashboard_context = f"<dashboards_context>Dashboards the user is looking at:\n{chr(10).join(dashboard_contexts)}\n</dashboards_context>"
```
```suggestion
                dashboard_context = f"<dashboards_context>\nDashboards the user is looking at:\n{chr(10).join(dashboard_contexts)}\n</dashboards_context>"
```

skoob13 commented on ee/hogai/graph/root/nodes.py:
```diff
@@ -52,7 +58,163 @@
+
+            if insights_results:
+                joined_results = "\n\n".join(insights_results)
+                insights_context = f"<standalone_insights_context>Insights the user is looking at:\n{joined_results}\n</standalone_insights_context>"
```
```suggestion
                insights_context = f"<standalone_insights_context>\nInsights the user is looking at:\n{joined_results}\n</standalone_insights_context>"
```

skoob13 commented on ee/hogai/graph/root/nodes.py:
```diff
@@ -52,7 +58,163 @@
+                        if insight_results:
+                            # Extract just the results part for inline display
+                            lines = insight_results.split("\n")
+                            results_start = next((i for i, line in enumerate(lines) if line.startswith("Results:")), -1)
```
This can easily break. I would just refactor the original code, so it returns results without the example prompts. The prompt injection helps the model to understand how data is structured, but if it works well without it, don't include it.

skoob13 commented on frontend/src/scenes/max/maxTypes.ts:
```diff
@@ -0,0 +1,53 @@
+}
+
+export interface MultiInsightContextContainer {
+    [insightKey: string]: InsightContextForMax
```
Do we have any reason to store dashboards and insights as maps? Should it be an array?

kappa90 replied:
Moved to arrays, instead of maps.

skoob13 reviewed (changes requested):
The main problem is that I couldn't make it run. When I send POST/conversations, it just throws 400 errors on the `ui_context` validation. Reviewed ~50% for now.

Larger feedback:
- I would expect I could pick a single insight on the page. You can't do it right now, and it's a little confusing.
- On some screens, it makes sense to always attach the context, e.g. when viewing a single insight. (Maybe not when editing it)

Commit:
fix: schema

Commit:
fix: remove useless types

skoob13 commented on frontend/src/scenes/max/Thread.tsx:
```diff
@@ -168,6 +169,13 @@ function MessageGroup({ messages, isFinal: isFinalGroup }: MessageGroupProps): J
                                 type="human"
                                 boxClassName={message.status === 'error' ? 'border-danger' : undefined}
                             >
+                                {message.ui_context && Object.keys(message.ui_context).length > 0 && (
```
### UI bugs & suggestions

**Message context**

It takes a lot of space. Can we make it one line and output "3 more" in a tooltip? I am honestly not sure it's helpful in any way, so I would either hide it or have a minor element clicking/hovering on which you can find the attached context.

Also, a sent human message has a layout shift since the ui_context is not attached until it gets data from the backend.

<img width="458" alt="Screenshot 2025-06-05 at 19 21 18" src="https://github.com/user-attachments/assets/4a2b59bf-e6d5-480d-ace4-3187824c9b12" />

**UI gets cluttered**

Can we think of a better way to have the context button? It's not necessarily in this PR, but the context button has added a new panel on top, which might be part of an existing text area.

<img width="490" alt="Screenshot 2025-06-05 at 19 28 03" src="https://github.com/user-attachments/assets/36afe28a-704b-43d8-9729-c9872830b9b1" />

**Layout of attached context**

I think they should all be contained in the same container, so the context button is not "isolated" from the context content. I attached two elements in the screenshot; the context button is always in the first line. I would expect it to be together with the first badge.

<img width="481" alt="Screenshot 2025-06-05 at 19 28 39" src="https://github.com/user-attachments/assets/645af685-75df-4727-b5ac-7f7e61190a46" />

**Context badges contrast**

The dark theme is +—okay, but the light theme is hard to read. Can we change the styling of badges so that they're easier to read?

<img width="481" alt="Screenshot 2025-06-05 at 19 28 39" src="https://github.com/user-attachments/assets/da15cce2-2bc4-4677-8510-9f84c55dbfd2" />

<img width="491" alt="Screenshot 2025-06-05 at 19 28 28" src="https://github.com/user-attachments/assets/e8020500-e896-4e2f-8481-834149ac0c40" />

Commit:
fix: github review

Commit:
fix: UI/UX fixes after review

Commit:
fix: review

Commit:
fix: add insights/dashboards to "in view" tab

Commit:
Merge remote-tracking branch 'origin/master' into max/frontend-context

Commit:
fix: products.tsx

kappa90 commented on frontend/src/scenes/max/maxTypes.ts:
```diff
@@ -0,0 +1,53 @@
+
+    query: Node // The actual query node, e.g., TrendsQuery, HogQLQuery
+
+    insight_type?: string
```
Removed.

kappa90 commented on ee/hogai/graph/query_executor/query_runner.py:
```diff
@@ -0,0 +1,197 @@
+                ExecutionMode.RECENT_CACHE_CALCULATE_ASYNC_IF_STALE
+                if not settings.TEST
+                else ExecutionMode.CALCULATE_BLOCKING_ALWAYS
+            )
```
Added a `ReasoningMessage`. For dashboards, we refresh the insights on the frontend before sending them as context. For single insights, we might still have problems, but that also applies to normal insight generation.

kappa90 commented on ee/hogai/graph/root/prompts.py:
```diff
@@ -48,6 +48,18 @@
+{{{ui_context_insights}}}
+{{{ui_context_navigation}}}
+
+</user_context>
```
Done.

kappa90 commented on ee/hogai/graph/root/prompts.py:
```diff
@@ -48,6 +48,18 @@
+
+{{{ui_context_dashboard}}}
+{{{ui_context_insights}}}
+{{{ui_context_navigation}}}
```
Grouped variables together.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

kappa90 commented:
@skoob13 I took all your invaluable feedback and fixed (hopefully) everything, bugs, code, and UI/UX :D

Notable changes / fixes:
- Added a `ReasoningMessage` when sending context, saying that we're "Calculating insights", to tell the user we might be loading the context queries before responding.
- Sending a dashboard as context now applies the dashboard filters to all the insights when running the queries. Sending an insight with overridden filters or variables also applies the right filters now.
- The user context root prompt is now only sent when context is available (it was always sent before)
- Removed navigation context as it doesn't add much value
- When opening a single insight (pathname: `/insight/:id`), the insight is automatically added to context (you can remove it from the UI)
- The context button and tags are now within the chat input, with a much cleaner UI. Everything is in one line and if there are too many tags, a [+{number}] tag appears, which can show extra tags in a tooltip.
![image](https://github.com/user-attachments/assets/0b63bf6c-adaf-4f69-9073-998119e9da91)
![image](https://github.com/user-attachments/assets/49245798-5e00-43e7-924c-a2b79f3a069c)
![image](https://github.com/user-attachments/assets/0963fb51-a584-425b-b18c-6748504907d0)

- Context sent in past messages now appears as a small summary above a human message, with more information on hover
![image](https://github.com/user-attachments/assets/2b530944-ac05-40db-bcce-d9443f4c61f5)
![image](https://github.com/user-attachments/assets/26a64f44-af24-4cc2-b240-3f1a66dbb043)

- The "Max AI" tab in the TaxonomicPopover has been renamed to "In this page". This now shows as options "Current Page", as well as the current dashboard and any insight in view. Only "Current Page" has an additional explanation Popover.
![image](https://github.com/user-attachments/assets/43302174-97bc-40e6-b5b7-01a6b02bce45)


Future PRs:
- Filter out from the TaxonomyPopover insight types that we don't support
- Handle insights that load for more than 180s and might lead to closing the HTTP stream

Note: this might break local past conversations, since the ui_context schema changed

skoob13 commented on ee/hogai/assistant.py:
```diff
@@ -239,9 +242,22 @@ def _stream(self) -> Generator[str, None, None]:
     def _initial_state(self) -> AssistantState:
         if self._latest_message and self._mode == AssistantMode.ASSISTANT:
-            return AssistantState(messages=[self._latest_message], start_id=self._latest_message.id)
+            # Add ui_context to the message if available
+            if self._ui_context:
+                message_with_ui_context = self._latest_message.model_copy(update={"ui_context": self._ui_context})
+                return AssistantState(
+                    messages=[message_with_ui_context],
+                    start_id=message_with_ui_context.id,
+                )
+            else:
+                return AssistantState(
+                    messages=[self._latest_message],
+                    start_id=self._latest_message.id,
+                )
```
Don't you do the same in the conversation serializer? The latest message should already have context since you set it in the serializer.

skoob13 commented on ee/hogai/graph/root/nodes.py:
```diff
@@ -89,12 +252,16 @@ def run(self, state: AssistantState, config: RunnableConfig) -> PartialAssistant
         utc_now = datetime.datetime.now(datetime.UTC)
         project_now = utc_now.astimezone(self._team.timezone_info)
 
+        ui_context = self._get_ui_context(config)
```
You previously mentioned keeping the UI context in messages to preserve it in history, but it is not used in any way. Should we use it?

skoob13 reviewed (commented):
Changes look good. Left a few general comments.

## Bugs

The insights are still duplicated when I choose the "Current page".

<img width="539" alt="Screenshot 2025-06-11 at 13 16 45" src="https://github.com/user-attachments/assets/b78fc18a-6b2f-44b1-89fa-d87846f7e8ac" />

## UI

This looks good.

<img width="309" alt="Screenshot 2025-06-11 at 13 04 59" src="https://github.com/user-attachments/assets/bacfde86-3b06-4d65-b787-22e6e921e1cb" />

For the attached context in the input, I'd probably keep the original (with wrapping) as it's much easier to work with the context. Maybe @Twixes has some ideas.

<img width="524" alt="Screenshot 2025-06-11 at 13 04 53" src="https://github.com/user-attachments/assets/cb1b4de4-5bdd-4fd2-845c-1aa89342de1c" />

Commit:
Merge remote-tracking branch 'origin/master' into max/frontend-context

Commit:
fix: review

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

Commit:
fix: tests

Commit:
fix: tests

skoob13 committed:
fix: shared function for ui_context lookup

skoob13 committed:
fix: context formatting

Commit:
Merge remote-tracking branch 'origin/master' into max/frontend-context

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

skoob13 committed:
chore: context understanding

skoob13 committed:
fix: deduplicate data

skoob13 committed:
Merge branch 'max/frontend-context' of github.com:PostHog/posthog into max/frontend-context

skoob13 commented:
Still can't always force the root to analyze the context with a simple prompt:

<img width="541" alt="Screenshot 2025-06-12 at 14 30 58" src="https://github.com/user-attachments/assets/7ce98bbf-3036-4e8a-a26e-0deaabe75546" />

After I had pushed changes, it became slightly more accurate:

<img width="511" alt="Screenshot 2025-06-12 at 14 30 46" src="https://github.com/user-attachments/assets/e2f04656-acfb-4426-b5e6-9fcd047b44da" />

We can fix it later.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Commit:
fix: mypy

Commit:
Merge remote-tracking branch 'origin/master' into max/frontend-context

Twixes committed:
Fix typing of _run_and_format_insight()

Commit:
fix: python types

Commit:
Merge remote-tracking branch 'origin/max/frontend-context' into max/frontend-context

Commit:
fix: types

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

kappa90 commented on ee/hogai/graph/query_executor/query_executor.py:
```diff
@@ -0,0 +1,284 @@
+        elif isinstance(query, HogQLQuery):
+            # Cast to AssistantHogQLQuery for consistent formatting
+            assistant_hogql_query = cast(AssistantHogQLQuery, query)
+            return SQLResultsFormatter(assistant_hogql_query, response["results"], response["columns"]).format()
```
Fixed, I also had forgotten about this file 😅

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

kappa90 committed:
Merge branch 'master' into max/frontend-context

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)
